### PR TITLE
Vulkan

### PR DESCRIPTION
**Pull request by vlj**
This PR moves program state creation into program_state_cache which should avoid a lot of creation/destruction. It also factorises sampler out of texture and  use a more explicit ctor for framebuffer.

### CODE REVIEW

File: rpcs3/Emu/RSX/VK/VKFormats.h [hunk 1]
```diff
@@ -14,4 +14,5 @@ namespace vk
 	VkSamplerAddressMode vk_wrap_mode(u32 gcm_wrap);
 	float max_aniso(u32 gcm_aniso);
 	VkComponentMapping get_component_mapping(u32 format, u8 swizzle_mask);
+	VkPrimitiveTopology get_appropriate_topology(rsx::primitive_type& mode, bool &requires_modification);
 }
```
Comment (tambry) [inline rpcs3/Emu/RSX/VK/VKFormats.h]: Missing newline.

File: rpcs3/Emu/RSX/VK/VKProgramPipeline.cpp [hunk 2]
```diff
@@ -502,28 +97,11 @@ namespace vk
...
-
-				if (pstate.pipeline_cache)
-					vkDestroyPipelineCache((*device), pstate.pipeline_cache, nullptr);
-			}
-
-			memset(&pstate, 0, sizeof pstate);
-			device = nullptr;
-		}
 	}
 }
```
Comment (tambry) [inline rpcs3/Emu/RSX/VK/VKProgramPipeline.cpp]: Same.

**Comment by kd-11** [comment]
I like the program state optimization. The sampler changes should be fine as long as sampler creation cost is negligible. Its a little odd that we go to all the trouble to avoid recreating psos but then make it so that we recreate samplers. If sampler creation cost is close to nil this isn't an issue though.

**Comment by vlj** [comment]
Fixed newline

@kd-11 I didn't notice performance drop with this PR. I'm thinking to put every sampler in an unordered_map anyway and store image with their image_view in texture cache.

**Pull request merged**